File: app/src/main/java/com/admin/theapp/view/HotelDetailsActivity.java
package com.admin.theapp.view;

import android.arch.lifecycle.Observer;
import android.graphics.drawable.BitmapDrawable;
import android.os.Bundle;
import android.support.annotation.NonNull;
import android.support.annotation.Nullable;
import android.widget.ImageView;
import android.widget.TextView;
import android.widget.Toast;

import com.admin.theapp.R;
import com.admin.theapp.base.BaseActivity;
import com.admin.theapp.ui.widget.StarsView;
import com.admin.theapp.viewmodel.HotelDetailsViewModel;
import com.theapp.entities.HotelModel;

import butterknife.BindView;

public class HotelDetailsActivity extends BaseActivity<HotelDetailsViewModel> {

    private static final int DEFAULT_ACTIVITY_RESULT_HOTEL_ID = -1;

    @BindView(R.id.hotel_details_image)
    ImageView image;
    @BindView(R.id.hotel_details_name)
    TextView  name;
    @BindView(R.id.hotel_details_address)
    TextView  address;
    @BindView(R.id.hotel_details_stars)
    StarsView stars;
    @BindView(R.id.hotel_details_distance)
    TextView  distance;
    @BindView(R.id.hotel_details_suites_availability)
    TextView  suites_availability;
    @BindView(R.id.hotel_details_lat)
    TextView  lat;
    @BindView(R.id.hotel_details_lon)
    TextView  lon;

    @NonNull
    private final Observer<HotelModel> hotelModelObserver = hotelModel -> {
        if (hotelModel != null) {
            name.setText(hotelModel.getName());
            address.setText(hotelModel.getAddress());
            stars.setStars(hotelModel.getStars());
            distance.setText(String.valueOf(hotelModel.getDistance()));
            suites_availability.setText(hotelModel.getSuitesAvailability());
            lat.setText(String.valueOf(hotelModel.getLat()));
            lon.setText(String.valueOf(hotelModel.getLon()));
            if (hotelModel.getImageName() != null) {
                viewModel.getImageByName(hotelModel.getImageName());
            }
        }
    };

    @NonNull
    private final Observer<BitmapDrawable> hotelImageObserver = img -> image.setImageDrawable(img);

    @Override
    protected void onCreate(@Nullable Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        getSupportActionBar().setDisplayHomeAsUpEnabled(true);
        final long hotelId = getIntent().getLongExtra(HotelsActivity.HOTEL_DETAILS_ACTIVITY_EXTRA, DEFAULT_ACTIVITY_RESULT_HOTEL_ID);
        if (hotelId == DEFAULT_ACTIVITY_RESULT_HOTEL_ID) {
            Toast.makeText(this, R.string.hotel_does_not_exist_text, Toast.LENGTH_SHORT).show();
            this.finish();
        }
        viewModel.retrieveHotelModel(hotelId);
        viewModel.getHotelModel().observe(this, hotelModelObserver);
        viewModel.getHotelImage().observe(this, hotelImageObserver);
    }

    @Override
    protected int getLayoutRes() {
        return R.layout.hotel_details;
    }

    @NonNull
    @Override
    protected Class<HotelDetailsViewModel> getViewModelClass() {
        return HotelDetailsViewModel.class;
    }
}

File: app/src/main/java/com/admin/theapp/view/HotelDetailsActivity.kt
package com.admin.theapp.view

import android.arch.lifecycle.Observer
import android.content.Intent
import android.graphics.Bitmap
import android.graphics.drawable.BitmapDrawable
import android.net.Uri
import android.os.Bundle
import com.admin.theapp.R
import com.admin.theapp.base.BaseActivity
import com.admin.theapp.utils.Constants
import com.admin.theapp.utils.toast
import com.admin.theapp.viewmodel.HotelDetailsViewModel
import com.theapp.entities.HotelModel
import kotlinx.android.synthetic.main.hotel_details.*

class HotelDetailsActivity : BaseActivity<HotelDetailsViewModel>() {
    override val vmClass = HotelDetailsViewModel::class.java
    override val layoutRes = R.layout.hotel_details

    private val hotelModelObserver = Observer<HotelModel>({
        it ?: return@Observer

        hotel_details_name.text = it.name
        hotel_details_address.text = it.address
        hotel_details_stars.setStars(it.stars)
        hotel_details_distance.text = it.distance.toString()
        hotel_details_suites_availability.text = it.suitesAvailability
        it.imageName?.let { viewModel.getImageByName(it) }
        viewModel.getMapThumbnail(it.lat, it.lon)
        hotel_details_on_map_button.setOnClickListener { _ ->
            intent = Intent(Intent.ACTION_VIEW, Uri.parse(getString(
                    R.string.map_query,
                    it.lat.toString(),
                    it.lon.toString(),
                    it.name.replace(' ', '+')
            )))
            if (intent.resolveActivity(packageManager) != null) startActivity(intent)
        }
    })

    private val hotelImageObserver = Observer<BitmapDrawable>({ hotel_details_image.setImageDrawable(it) })

    private val mapImageObserver = Observer<Bitmap>({ hotel_details_on_map_button.setImageBitmap(it) })

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        supportActionBar?.setDisplayHomeAsUpEnabled(true)
        val hotelId = intent.getLongExtra(Constants.EXTRA_HOTEL_DETAILS_ACTIVITY, DEFAULT_ACTIVITY_RESULT_HOTEL_ID)
        if (hotelId == DEFAULT_ACTIVITY_RESULT_HOTEL_ID) {
            toast(R.string.hotel_does_not_exist_text)
            finish()
        }
        viewModel.retrieveHotelModel(hotelId)
        viewModel.hotelModel.observe(this, hotelModelObserver)
        viewModel.hotelImage.observe(this, hotelImageObserver)
        viewModel.mapImage.observe(this, mapImageObserver)
    }

    private companion object {
        const val DEFAULT_ACTIVITY_RESULT_HOTEL_ID: Long = -1
    }
}
